Qi/online catalog (SIMBAD)

panchyo0 opened the pull request:
related issue #1243
1. `search radius` range is 0 - 90 deg according simbad.
2. [In Simbad, the coordinate system parameter is never interpreted. All coordinates MUST be expressed in the ICRS. coordinate system ](http://simbad.u-strasbg.fr/simbad/tap/help/adqlHelp.html#geometricalFunctions), so only ICRS available for SIMBAD as `center coordinate` system.
3. `center coordinate` will auto update once user frozen curser at image viewer.

panchyo0 committed:
refactor catalog profile store

panchyo0 committed:
query simbad by radius, coord and max number; init UI

panchyo0 committed:
minor bug fixed

panchyo0 committed:
clear up

panchyo0 committed:
fixed coord system convert bug, add sort, minor UI update

panchyo0 committed:
add filters

panchyo0 committed:
minor bugs fixed

panchyo0 committed:
add object search box, minor ui update

panchyo0 committed:
added object coord search, fixed source selection bug with filter

panchyo0 committed:
add show selected with filter and sort, fixed scatter plot update bug

panchyo0 committed:
merge dev

panchyo0 committed:
minor UI update

panchyo0 committed:
code clear up

panchyo0 committed:
reformate with Prettier

panchyo0 committed:
minor bugs fixed

panchyo0 committed:
add hyper link for simbad main id

panchyo0 committed:
reformate

panchyo0 committed:
minor bug fixed

kswang1029 commented:
I tested this PR and found the following issues/suggestions/comments so far:

- when rendering the retrieved catalog as an image overlay, it seems always to have position offset and the sources are not matched to the features in the image. The following example is the famous M51 galaxy, but the returned source (setting the max number of objects = 1) appears not correct. 

![Screen Shot 2021-09-13 at 2 06 19 PM](https://user-images.githubusercontent.com/20819712/133032034-9626a957-e250-44fd-a373-e4fa4daca646.png)

- Can we use a button named as "Resolve" to the right of the text field? It would be more intuitive than the current implementation I think. We might consider to use an icon next to the button to show if the source is found (with v) or not (with x).

- for the "Search Radius" entry, can we support unit conversion too? That is, if I enter 1 with unit deg and switch to arcmin via the dropdown, the value changes to 60.

- for the "Search Radius" entry, it is better to keep the unit persist as it was after pressing the "set to image" button. Now it is always be degree.

- for the "center coordinates", we should also support the sexagesimal format. 

- for the "center coordinates", the coordinate reference should be better aligned with the image. That is, if the image is in GAL, then the default should be GAL for better UX. 

- could you explain a bit on how the "search radius" is derived when "set to image" button is clicked? (maybe it should be "set to viewer" to be more precise?)

- the long title is not displayed well with the catalog setting dialog. it is a global issue when the title is long...

![Screen Shot 2021-09-13 at 2 16 25 PM](https://user-images.githubusercontent.com/20819712/133033104-2ab3dcc4-8382-4cec-9180-90faca4d76b3.png)

- I suggest we set the following as the default displayed columns from SIMBAD.

![Screen Shot 2021-09-13 at 2 32 14 PM](https://user-images.githubusercontent.com/20819712/133034971-a674b3bb-d917-4658-bdab-c981da87dda2.png)

veggiesaurus commented on src/components/Dialogs/CatalogQueryDialog/CatalogOnlineQueryDialogComponent.tsx:
```diff
@@ -0,0 +1,305 @@
+export class CatalogQueryDialogComponent extends React.Component {
+    private static readonly DefaultWidth = 550;
+    private static readonly DefaultHeight = 500;
+    private static readonly DBMap = new Map<CatalogDatabase, {type: CatalogSystemType[]; prefix: string}>([
```
please move anything interacting with an external API via HTTP to a separate `Service` class

veggiesaurus commented on src/components/Dialogs/CatalogQueryDialog/CatalogOnlineQueryDialogComponent.tsx:
```diff
@@ -0,0 +1,305 @@
+        const baseUrl = CatalogQueryDialogComponent.DBMap.get(configStore.catalogDB).prefix;
+        const query = `SELECT Top ${configStore.maxObject} * FROM basic WHERE CONTAINS(POINT('ICRS',ra,dec),CIRCLE('ICRS',${configStore.centerCoord.x},${configStore.centerCoord.y},${configStore.radiusInDeg}))=1 AND ra IS NOT NULL AND dec IS NOT NULL`;
+        configStore.setQueryStatus(true);
+        AppStore.Instance.appendOnlineCatalog(baseUrl, query, this.cancelTokenSource)
```
`appendOnlineCatalog` should be moved to a service class

veggiesaurus commented on src/models/Processed.ts:
```diff
@@ -169,3 +169,60 @@ export class ProtobufProcessing {
     }
 }
+
+enum SIMBADTYPE {
```
`SimbadType`

veggiesaurus commented on src/models/Processed.ts:
```diff
@@ -169,3 +169,60 @@ export class ProtobufProcessing {
+    DOUBLE = "DOUBLE"
+}
+
+export class APIProcessing {
```
separate file

veggiesaurus commented on src/services/ApiService.ts:
```diff
@@ -505,4 +505,13 @@ export class ApiService {
         }
     };
+
+    public getSimbad = (apiAddress: string, query: string, cancelTokenSource?: CancelTokenSource): Promise<AxiosResponse<any>> => {
```
move into separate class. `ApiService` is forinteracting with the CARTA backend/server API. We can rename this class if you want

veggiesaurus commented on src/services/ApiService.ts:
```diff
@@ -505,4 +505,13 @@ export class ApiService {
+
+    public getSimbad = (apiAddress: string, query: string, cancelTokenSource?: CancelTokenSource): Promise<AxiosResponse<any>> => {
+        if (cancelTokenSource) {
+            return axios.get(`${apiAddress}${query}`, {
```
you can set the base URL in a new axios instance

veggiesaurus commented on src/stores/AppStore.ts:
```diff
@@ -873,6 +848,97 @@ export class AppStore {
         });
     };
 
+    // Online Catalog Query
```
you can move some of this to the newly created service class. 

Can you not just use `await` to simplify this?

veggiesaurus commented on src/stores/CatalogOnlineQueryProfileStore.ts:
```diff
@@ -0,0 +1,222 @@
+        return this.loadingData;
+    }
+
+    @computed get maxRows(): number {
```
what's the point of having a `@computed` wrapper around an `@observable`? is this just so that it adheres to the parent class? If so, I'd make the `@observable` values private

veggiesaurus commented on src/stores/CatalogOnlineQueryProfileStore.ts:
```diff
@@ -0,0 +1,222 @@
+    }
+
+    @action initSortedIndexMap() {
+        this.sortedIndexMap = [];
```
shouldn't `sortedIndexMap` be a `Map<>`?

panchyo0 replied:
the order of this array is represented the sort order. And the value is the index in the original catalog data 😄. Should we rename it to `sortedIndexArray`

veggiesaurus commented on src/utilities/parsing.ts:
```diff
@@ -61,7 +61,11 @@ export function findDeep(obj: any, pred: (obj: any) => boolean) {
 // parsing filter string for TableComponent filter function
 function getNumberFromFilterString(filterString: string): number {
-    return Number(filterString.replace(/[^0-9.+-.]+/g, ""));
+    const n = filterString.replace(/[^0-9.+-.]+/g, "");
```
can we not turn this into a single regex with the comparison operator? @confluence is our regex expert here :laughing:

panchyo0 replied:
This regex will return the numerical part from the filter string. Not sure what we should do at here 🤔

veggiesaurus reviewed (changes requested):
making great progress :sunglasses:

panchyo0 committed:
add dist columns, order by dist as default. object UI update

panchyo0 committed:
search radius update

panchyo0 committed:
update center coordinates function, display text overflow as ellipsis for floating widget title

panchyo0 committed:
refactor code

panchyo0 committed:
reformate code

panchyo0 committed:
fix pixel convert, reset object when center coordinate updated

panchyo0 commented:
Hi @kswang1029, I fixed those issues, Please check again, thanks
> when rendering the retrieved catalog as an image overlay, it seems always to have position offset and the sources are not matched to the features in the image. The following example is the famous M51 galaxy, but the returned source (setting the max number of objects = 1) appears not correct.

I think the offset comes from the `search radius`, SIMBAD does not return the data according the distance to the source. To solve this, I added another column `dist` as an extra column to the returned data and set the default order according `dist`

> ould you explain a bit on how the "search radius" is derived when "set to image" button is clicked? (maybe it should be "set to viewer" to be more precise?)

when user click the `set to viewer` button, the `center coordinate` will auto update to the current view center. Then I get the pixel coordinates for X1(xmax, ymax) and X2 (xmin, ymin), and convert X1 and X2 to ICRS coordinates (degree). After that I will calculate the distance between X1 and X2. Then the `search radius = the distance between X1 and X2 / 2` Not sure if this is the right way to do it 🙈

kswang1029 commented:
> when user click the `set to viewer` button, the `center coordinate` will auto update to the current view center. Then I get the pixel coordinates for X1(xmax, ymax) and X2 (xmin, ymin), and convert X1 and X2 to ICRS coordinates (degree). After that I will calculate the distance between X1 and X2. Then the `search radius = the distance between X1 and X2 / 2` Not sure if this is the right way to do it 🙈

this sounds sensible I think. 🙂 As the image field might have a rotation or have significant projection distortion, I think we should also check the other diagonal distance (xmin, ymax) (xmax, ymin) then take their max. Ideally (probably overkill...), we should check
<img width="560" alt="Screen Shot 2021-09-16 at 1 03 09 PM" src="https://user-images.githubusercontent.com/20819712/133552593-c3153302-d918-4148-a6eb-19519a3fa704.png">

panchyo0 committed:
update radius method

kswang1029 commented:
Few more feedback:

- not all the entries in the "xxx_bibcode" columns are hyperlink, e.g.,
![Screen Shot 2021-09-24 at 10 25 51 AM](https://user-images.githubusercontent.com/20819712/134609173-be89981c-e93d-4313-b637-a4fb611c68ed.png)

column index of xxx_bibcode: [1, 11, 18, 24, 30, 42, 52] 

- usability improvement: the "dist" column is now displayed in degree which may not be that friendly to some users. It would be better if we use arcsec as the unit, same as SIMBAD, so that users feel familiar with.

- usability improvement: at the moment ra and dec are displayed in degree. It would be good to construct a _new_ column as string to show the the coordinate (ra, dec) in sexagesimal format (note sorting in the column will become meaning less but it is fine).

- I tried to use the distance measuring tool to test the search radius estimate. In the following example, the search radius is estimated as 8.689197478226912 deg but from the distance measuring tool, the distance from bottom-left to top-right corner (roughly as I cannot click the true corners due to the blockage of the tool bar and cursor info elements) is ~25 deg. So the radius should be ~12 deg-ish. Any idea?

![Screen Shot 2021-09-24 at 11 01 53 AM](https://user-images.githubusercontent.com/20819712/134611930-79fe7d51-cfc8-4d73-811c-079250d7ec72.png)


ps sorting and filtering work fine 👍

panchyo0 committed:
convert dist to arcsec; add sexagesimal formate for ra and dec; add hyperlink to all bibcode

veggiesaurus commented on src/services/CatalogApiService.ts:
```diff
@@ -0,0 +1,109 @@
+}
+
+export class CatalogApiService {
+    public static readonly SimBadHyperLink: {bibcode: string; mainId: string} = {bibcode: "https://ui.adsabs.harvard.edu/abs/", mainId: "https://simbad.u-strasbg.fr/simbad/sim-id?Ident="};
```
lower-case `B`

veggiesaurus commented on src/services/CatalogApiService.ts:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    constructor() {
+        // makeObservable(this);
```
remove comment

veggiesaurus commented on src/services/CatalogApiService.ts:
```diff
@@ -0,0 +1,109 @@
+        }
+
+        const fileId = AppStore.Instance.catalogNextFileId;
+        const dataSize = await this.getSimbad(query)
```
confused as to why you are awaiting _and_ using `.then`. 

why not do 

```typescript
let dataSize: number;
try {
const res = await this.getSimbad(query);
if (...) {
   dataSize = response.data?.data?.length;
}
}
catch(err) {
...
}
```

veggiesaurus commented on src/services/CatalogApiService.ts:
```diff
@@ -0,0 +1,109 @@
+        });
+    }
+
+    public getSimbad = (query: string): Promise<AxiosResponse<any>> => {
```
rename `getSimbadCatalog` or `getSimbadResponse`

panchyo0 committed:
fixed radius calculation bug

panchyo0 commented:
@kswang1029 could you check again about the `search radius `  bug?

kswang1029 commented:
> @kswang1029 could you check again about the `search radius ` bug?

Yes, now the search radius should be correct. 👍 

Minor issues:
<img width="1446" alt="Screen Shot 2021-09-28 at 10 33 37 AM" src="https://user-images.githubusercontent.com/20819712/135013634-a28ea6b0-84bd-4759-b201-ff445604c3aa.png">

panchyo0 committed:
minor ui issues fixed

panchyo0 committed:
merge dev

panchyo0 commented:
@kswang1029  I have fixed those minor issues, please check.

kswang1029 reviewed (approved):
Nice work 👍

veggiesaurus commented on src/components/Dialogs/CatalogQueryDialog/CatalogOnlineQueryDialogComponent.tsx:
```diff
@@ -0,0 +1,376 @@
+        }
+    };
+
+    private handleCenterWCSXChange = ev => {
```
rename this and other functions `Wcs` instead of `WCS`

panchyo0 replied:
there are some other functions like `getFormattedWCSPoint, getPixelValueFromWCS, isWCSStringFormatValid` in utilities, should I rename all of them? 🤔

veggiesaurus commented on src/components/Shared/ClearableNumericInput/ClearableNumericInputComponent.tsx:
```diff
@@ -77,7 +78,7 @@ export class ClearableNumericInputComponent extends React.Component<ClearableNum
         let value = this.props.displayExponential && !this.isFocused ? toExponential(Number(this.props.value), 3) : this.props.value;
         return (
-            <FormGroup className={this.props.className} label={this.props.label} inline={true} disabled={this.props.disabled}>
+            <FormGroup className={this.props.className} label={this.props.label} inline={this.props.inline === undefined ? true : false} disabled={this.props.disabled}>
```
` ? true : false` is unnecessary, isn't it?

panchyo0 replied:
the default value is `false or undefined` from blueprintjs, in our case we need the default to be `true` 🙈 this is a one line solution, we also can add `inline = true` to all `ClearableNumericInputProps`, not sure which one is better?

veggiesaurus replied:
`this.props.inline === undefined` returns true or false already

veggiesaurus commented on src/models/CatalogApiProcessed.ts:
```diff
@@ -0,0 +1,106 @@
+    DOUBLE = "DOUBLE"
+}
+
+export class APIProcessing {
```
please match the file name

veggiesaurus commented on src/services/CatalogApiService.ts:
```diff
@@ -0,0 +1,106 @@
+                        dataSize: response.data?.data?.length,
+                        directory: ""
+                    };
+                    let catalogWidgetId = AppStore.Instance.updateCatalogProfile(fileId, frame);
```
use `const appStore = AppStore.Instance`

veggiesaurus commented on src/services/CatalogApiService.ts:
```diff
@@ -0,0 +1,106 @@
+
+    // Online Catalog Query
+    public appendOnlineCatalog = async (query: string): Promise<number> => {
+        const frame = AppStore.Instance.activeFrame;
```
`const appStore = ...`

panchyo0 committed:
clear up

kswang1029 reviewed (approved):
I have also tested the sorting and filtering functions and compared with the backend sorting and filtering functions. All looking good. 

one thing I noticed is the "dist" calculation has ~0.01-0.1 arcsec difference comparing to the downloaded votable. It would be good to try to make the dist calculation as precise as possible. This minor issue is not critical though.

veggiesaurus commented on src/components/Dialogs/CatalogQueryDialog/CatalogOnlineQueryDialogComponent.tsx:
```diff
@@ -0,0 +1,376 @@
+        let resultStr = undefined;
+        const configStore = CatalogOnlineQueryConfigStore.Instance;
+        if (configStore.isQuerying || configStore.isObjectQuerying) {
+            resultStr = `Querying ${configStore.catalogDB}`;
```
replace with just `return` instead of assigning `resultStr`

veggiesaurus commented on src/components/Dialogs/CatalogQueryDialog/CatalogOnlineQueryDialogComponent.tsx:
```diff
@@ -0,0 +1,376 @@
+        const query = `SELECT basic.* FROM ident JOIN basic ON ident.oidref = basic.oid WHERE id = '${configStore.objectName}'`;
+        configStore.setObjectQueryStatus(true);
+        CatalogApiService.Instance.getSimbadCatalog(query)
+            .then(response => {
```
why don't you use await here instead?

veggiesaurus commented on src/models/CatalogApiProcessed.ts:
```diff
@@ -0,0 +1,106 @@
+    DOUBLE = "DOUBLE"
+}
+
+export class CatalogApiProcessing {
```
These don't belong belong under `models`. It's just a bunch of static functions. perhaps something in `utilities`?

I guess the same is true for the `ProtobufProcessing` class, so perhaps move that one to utilities too?

veggiesaurus commented on src/services/CatalogApiService.ts:
```diff
@@ -0,0 +1,107 @@
+                    const configStore = CatalogOnlineQueryConfigStore.Instance;
+                    const headers = CatalogApiProcessing.ProcessSimbadMetaData(response.data?.metadata);
+                    const columnData = CatalogApiProcessing.ProcessSimbadData(response.data?.data, headers);
+                    const coosy: CARTA.ICoosys = {system: configStore.coordsType};
```
typo?

panchyo0 committed:
code restyle

panchyo0 committed:
fixed conflicts

veggiesaurus committed:
Merge branch 'dev' into qi/issue_1243

veggiesaurus committed:
fixed compile error after merge

veggiesaurus commented on src/components/Shared/Tables/FilterableTableComponent.tsx:
```diff
@@ -162,37 +168,42 @@ export class FilterableTableComponent extends React.Component<FilterableTableCom
-            }
-        } else {
-            contents = "";
+        let rowIndex = index;
```
@panchyo0 here's the issue with #1593 ...see the handling of boolean columns in the previos commit has been removed by this PR

veggiesaurus commented on src/components/CatalogOverlay/CatalogOverlayComponent.tsx:
```diff
@@ -394,113 +388,10 @@ export class CatalogOverlayComponent extends React.Component<WidgetProps> {
-                if (dataType === CARTA.ColumnType.String) {
-                    filter.subString = value.filter;
-                    userFilters.push(filter);
-                } else if (dataType === CARTA.ColumnType.Bool) {
```
this has also now been removed @panchyo0, and breaks #1593
